src/rules/defineFlowType.js

```diff
@@ -3,87 +3,78 @@ const schema = [];
-
-      if (ref.identifier.name === ident.name) {
-        // use "__defineGeneric" since we don't have a reference to "escope.Variable"
-
-        globalScope.__defineGeneric(
-          ident.name,
-          globalScope.set,
-          globalScope.variables,
-        );
-        const variable = globalScope.set.get(ident.name);
-
-        variable.writeable = false;
-
-        // "through" contains all references whose definition cannot be found
-        // so we need to update references and remove the ones that were added
-        globalScope.through.splice(ii, 1);
-        ref.resolved = variable;
-        variable.references.push(ref);
+  const makeDefined = (variableName) => {
+    // Add the variable to the global scope
+    globalScope.through = globalScope.through.filter((ref) => {
+      if (ref.identifier.name === variableName) {
+        globalScope.variables.push({
+          name: variableName,
+          identifiers: [ref.identifier],
+          references: [ref],
+          defs: [],
+        });
+        return false;
       }
-    }
+      return true;
+    });
   };
 
   // NOTE: For future contributors, if you ever need to add support for a new identifier,
   // use `Identifier(node) {}` to find out which identifiers should be handled.
   return {
     ClassImplements(node) {
-      makeDefined(node.id);
+      makeDefined(node.id.name);
     },
     DeclareInterface(node) {
-      makeDefined(node.id);
+      makeDefined(node.id.name);
     },
     DeclareTypeAlias(node) {
-      makeDefined(node.id);
+      makeDefined(node.id.name);
     },
     EnumDeclaration(node) {
-      makeDefined(node.id);
+      makeDefined(node.id.name);
     },
     EnumDefaultedMember(node) {
-      makeDefined(node.id);
+      makeDefined(node.id.name);
     },
     EnumNumberMember(node) {
-      makeDefined(node.id);
+      makeDefined(node.id.name);
     },
     EnumStringMember(node) {
-      makeDefined(node.id);
+      makeDefined(node.id.name);
     },
     GenericTypeAnnotation(node) {
       if (node.id.type === 'Identifier') {
-        makeDefined(node.id);
+        makeDefined(node.id.name);
       } else if (node.id.type === 'QualifiedTypeIdentifier') {
         let qid;
 
         qid = node.id;
-        do {
+        while (qid.qualification) {
           qid = qid.qualification;
-        } while (qid.qualification);
+        }
 
-        makeDefined(qid);
+        makeDefined(qid.name);
       }
     },
 
     // Can be removed once https://github.com/babel/babel-eslint/pull/696 is published
     OpaqueType(node) {
       if (node.id.type === 'Identifier') {
-        makeDefined(node.id);
+        makeDefined(node.id.name);
       }
     },
-    Program() {
-      globalScope = context.getScope();
+    Program(node) {
+      const newGetScope = context.sourceCode.getScope;
+      if (newGetScope) {
+        globalScope = context.sourceCode.getScope(node);
+      } else {
+        globalScope = context.getScope();
+      }
```

Comment:
@Brianzchen `context.sourceCode` can be undefined in older versions of eslint so this code causes the following error: `TypeError: Cannot read properties of undefined (reading 'getScope')`.

I think you want something like:
```
      if (context.sourceCode != null && context.sourceCode.getScope != nul) {
        globalScope = context.sourceCode.getScope(node);
      } else {
        globalScope = context.getScope();
      }
```